### PR TITLE
Feature/45 display suggestions

### PR DESCRIPTION
**Pull request by rcffc**

### CODE REVIEW

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/recommender/NamedEntityLinker.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+import de.tudarmstadt.ukp.inception.kb.graph.KBHandle;
+import de.tudarmstadt.ukp.inception.kb.model.KnowledgeBase;
+import de.tudarmstadt.ukp.inception.recommendation.imls.conf.ClassifierConfiguration;
+import de.tudarmstadt.ukp.inception.recommendation.imls.core.classifier.Classifier;
+import de.tudarmstadt.ukp.inception.recommendation.imls.core.dataobjects.AnnotationObject;
+import de.tudarmstadt.ukp.inception.recommendation.imls.core.dataobjects.Offset;
+import de.tudarmstadt.ukp.inception.recommendation.imls.core.dataobjects.TokenObject;
+
+public class NamedEntityLinker
+
```
Comment (reckart) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/recommender/NamedEntityLinker.java]: Remove empty line.

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/recommender/NamedEntityLinker.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    extends Classifier<Object>
+{
+    private Logger log = LoggerFactory.getLogger(getClass());
```
Comment (reckart) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/recommender/NamedEntityLinker.java]: Logger should be final.

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/recommender/NamedEntityLinker.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+            nerAnnotations = (Set<AnnotationObject>) aModel;
+        }
+        else {
+            log.error("Expected model type: Set<TokenObject> - but was: [{}]",
+                aModel != null ? aModel.getClass() : aModel);
+        }
+    }
+
+    @Override
+    public void setUser(User user)
```
Comment (reckart) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/recommender/NamedEntityLinker.java]: Arguments should start with "a" and "this." is then not necessary.

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/recommender/NamedEntityLinker.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+            }
+            result.add(annotatedSentence);
+        }
+        return result;
+    }
+
+    private List<AnnotationObject> predictToken(TokenObject token)
+    {
+        List<KBHandle> handles = new ArrayList<>();
+        for (KnowledgeBase kb : kbService.getKnowledgeBases(project)) {
```
Comment (reckart) [inline inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/recommender/NamedEntityLinker.java]: This should respect the concept feature traits: kb and scope - just like in the ConceptFeatureEditor.

File: inception-imls-core/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/core/classifier/Classifier.java [hunk 5]
```diff
@@ -157,12 +164,15 @@ public Classifier(ClassifierConfiguration<C> conf)
...
                 i++;
             }
         }
         
         return predictionsWithMergedLabels;
     }
+
+    public abstract void setUser(User user);
+
+    public abstract void setProject(Project project);
```
Comment (reckart) [inline inception-imls-core/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/core/classifier/Classifier.java]: You can reduce the number of files affected by adding user and project as fields to the abstract class classifier and fully implement setters and getters. In your NamedEntityLinkingClassifier, you can then just call the getter.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/adapter/RecommendationSpanRenderer.java [hunk 6]
```diff
@@ -116,9 +121,10 @@ public void render(JCas aJcas, VDocument vdoc, AnnotatorState aState,
             // show only the confidence of the highest one
             for (AnnotationObject ao: token) {          
                 if (ao.getAnnotation() != null) {       
-                    if (isOverlapping(vspansWithoutRecommendations, ao.getOffset(), windowBegin)) {
-                        break;
-                    }
+//                    if (isOverlapping(vspansWithoutRecommendations, ao.getOffset(), windowBegin,
```
Comment (reckart) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/adapter/RecommendationSpanRenderer.java]: I have a strong feeling this should be properly restored before merging the PR...

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/adapter/RecommendationSpanRenderer.java [hunk 7]
```diff
@@ -173,7 +179,7 @@ public void render(JCas aJcas, VDocument vdoc, AnnotatorState aState,
                 // Create VID using the recommendation with the lowest recommendationId
                 AnnotationObject aPrediction = token.stream()
```
Comment (reckart) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/adapter/RecommendationSpanRenderer.java]: Never assign new values to argument variables (those starting with `a`).

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/adapter/RecommendationSpanRenderer.java [hunk 8]
```diff
@@ -173,7 +179,7 @@ public void render(JCas aJcas, VDocument vdoc, AnnotatorState aState,
...
                         .filter(p -> p.getAnnotation().equals(label))
-                        .max((p1, p2) -> Integer.compare(p1.getId(), p2.getId())).get();
+                        .max(Comparator.comparingInt(TokenObject::getId)).orElse(null);
```
Comment (reckart) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/adapter/RecommendationSpanRenderer.java]: Can you really handle the case where the prediction becomes null?!

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/adapter/RecommendationSpanRenderer.java [hunk 9]
```diff
@@ -183,8 +189,13 @@ public void render(JCas aJcas, VDocument vdoc, AnnotatorState aState,
...
 
                     if (first) {
-                        HashMap<String, String> featureAnnotation = new HashMap<String, String>();
-                        featureAnnotation.put(ao.getFeature(), ao.getAnnotation());
+                        AnnotationFeature feature = aAnnotationService
+                            .getFeature(ao.getFeature(), layer);
+                        String annotation = defaultString(aFsRegistry.getFeatureSupport(feature)
+                            .renderFeatureValue(feature, ao.getAnnotation()));
+
+                        HashMap<String, String> featureAnnotation = new HashMap<>();
```
Comment (reckart) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/adapter/RecommendationSpanRenderer.java]: Map<String, String> ... =

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/adapter/RecommendationSpanRenderer.java [hunk 10]
```diff
@@ -196,6 +207,9 @@ public void render(JCas aJcas, VDocument vdoc, AnnotatorState aState,
                     vdoc.add(new VComment(vid, VCommentType.INFO, ao.getClassifier()));
                     vdoc.add(new VComment(vid, VCommentType.INFO,
                             "Confidence: " + String.format("%.2f", ao.getConfidence())));
```
Comment (reckart) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/adapter/RecommendationSpanRenderer.java]: pull "Confidence: " into the format string.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/model/Predictions.java [hunk 11]
```diff
@@ -166,12 +166,9 @@ public Predictions(Project aProject, User aUser) {
         Type type = CasUtil.getType(aJcas.getCas(), aLayer.getName());
         List<AnnotationFS> existingAnnotations = CasUtil.selectCovered(aJcas.getCas(), 
                 type, aWindowBegin, aWindowEnd);
-        List<Integer> existingOffsets = existingAnnotations.stream().map(e -> e.getBegin())
-                .collect(Collectors.toList());
 
         return predictions.entrySet().stream()
                 .filter(f -> f.getKey().getDocumentName().equals(aDocumentName))
-                .filter(f -> !existingOffsets.contains(f.getKey().getOffset().getBeginCharacter()))
```
Comment (reckart) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/model/Predictions.java]: Same as commented out overlapping code before: handling overlapping annotations should be restored before merging the PR.

File: inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/feature/ConceptFeatureSupport.java [hunk 12]
```diff
@@ -159,9 +156,15 @@ public String renderFeatureValue(AnnotationFeature aFeature, AnnotationFS aFs,
...
     {
-        KBHandle kbInst = (KBHandle) aValue;
         FeatureStructure fs = selectByAddr(aJcas, FeatureStructure.class, aAddress);
-        setFeature(fs, aFeature, kbInst != null ? kbInst.getIdentifier() : null);
+        String kbIdentifier;
+        if (aValue instanceof KBHandle) {
+            kbIdentifier = ((KBHandle) aValue).getIdentifier();
+        }
+        else {
+            kbIdentifier = (String) aValue;
```
Comment (reckart) [inline inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/feature/ConceptFeatureSupport.java]: Better create a KBHandle for the identifier before calling this method and remove the instanceof here. Method should always be called with a KBHandle.

File: inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/recommender/NamedEntityLinker.java (hunk 4 above)
Comment (rcffc) [reply inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/recommender/NamedEntityLinker.java]: Adding ConceptFeatureTraits to inception-concept-linking will introduce a circular dependency to inception-ui-kb. I will create a new issue for that.

**Comment by reckart** [comment]
@rcffc dependencies need to be fixed (cf jenkins logs or build locally).

**Pull request merged**